### ISSUE TITLE
Blend factors and Ops should be a part of dynamic state.

### ISSUE THREAD

**Issue by sasmaster** [KhronosGroup/Vulkan-Docs#1172]
Isn't it weird that blend factors and ops are hardcoded into VkPipeline setup and are not part of Vulkan dynamic state? I mean,if I render A,B,C,D meshes each with different with different blend mode, Vulkan forces me to use different pipelines just for that per draw call?
I undersrtand that mobile devices might have performance issue with dynamic setup of blend modes, but for desktops an extension could be provided.  
For instance, I use the following blend modes extensively in rendering solutions I work on:
https://www.khronos.org/registry/OpenGL/extensions/NV/NV_blend_equation_advanced.txt 

Not just an alpha blending. So while in GL I would change the mode before every draw call,in Vulkan I must generate a pipeline per such a mode. I do see that blend constants are part of dynamic state, so why the rest of important blend settings are not there?

**Comment by Degerz** [comment]
It's probably because this will trigger shader recompilation which is expensive on many vendors and it especially applies to mobile devices like you mentioned ... 

I guess so far that only Nvidia and maybe AMD can support dynamic blend states in Vulkan ...

**Comment by TomOlson** [comment]
The WG discussed this this morning - we revisit the question of what state is dynamic pretty regularly. Unfortunately, there isn't a lot we can do to make blend state more dynamic with current hardware. As @Degerz said, hardware support for dynamic blend state isn't as common as you might think, so requiring it in the API means that you'd get submit-time recompilation on a significant slice of hardware. If apps actually use it, IHVs would feel like they have to add a layer of under-the-hood pipeline caching to (try to) hide the recompilation, and it would get ugly pretty fast.

So, Khronos isn't going to define an extension for this any time soon. Individual GPU vendors certainly could if they have the HW support for it, so you might talk to dev rel at the vendors you care about to see if they can do anything. 

We do appreciate the request - we are always glad to hear what developers want from the API, even when we can't do what they want in the short term. Blend state was already on our list of "things people wish were dynamic", but we'll add this issue to our internal tracker.

**Comment by sasmaster** [comment]
So far I can see you did release VK_EXT_extended_dynamic_state,so I hope blending modes will be added there one day as well  ;)

**Comment by libcg** [comment]
This would be useful for https://github.com/libcg/grvk, along with MSAA and polygon mode states.

**Comment by PJB3005** [comment]
Am I correct that this is now done as part of the new `VK_EXT_extended_dynamic_state3`?

**Comment by ShabbyX** [comment]
Yes indeed. Of course, it depends on an optional feature so it's only usable on hardware where blend state is truly dynamic.
